src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,157 @@
+        # available = standby_replay + standby_active
+        assert self.fs_map is not None
+        total = 0
+        daemon_names = {
+            d.name() for d in daemons
+        }
+        for sb in self.fs_map['standbys']:
+            full_name = "mds.{}".format(sb['name'])
+            if full_name in daemon_names:
+                total += 1
+        return total
+
+    def get_current_active_count(self, fs_name: str):
+        assert self.fs_map is not None
+        for fsys in self.fs_map.get('filesystems'):
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return len(fsys.get('mdsmap').get('in'))
+        return 0
+
+    def get_fs_name(self, index: int = 0):
+        assert self.fs_map is not None
+        fs = self.fs_map.get('filesystems')[index]
+        fs_name = fs.get('mdsmap').get('fs_name')
+        return fs_name
+
+    def verify_and_manage_mds_instance(self, fs_name: str):
+        assert self.fs_map is not None
+
+        daemons = self.get_daemons(fs_name)
+        standbys_required = self.get_required_standby_count(fs_name)
+        standbys_current = self.get_current_standby_count(daemons, fs_name)
+        active = self.get_current_active_count(fs_name)
+
+        self.log.info("standbys_required:{0}, standbys_current:{1}"
+                      .format(standbys_required, standbys_current))
+        if standbys_required is None:
+            return
+        total = standbys_current + active
+        if standbys_current > standbys_required:
+            # remove one mds at a time and wait for updated fs_map
+            self.log.info("killing {} standby mds ..."
+                          .format((standbys_current - standbys_required)))
+            total -= standbys_current - standbys_required
+        elif standbys_current < standbys_required:
+            # launch one mds at a time and wait for updated fs_map
+            self.log.info("spawning {} standby mds ..."
+                          .format((standbys_required - standbys_current)))
+            total += standbys_required - standbys_current
+
+        newspec = self.update_daemon_count(fs_name, total)
+        # newspec == None means that the placement count is already at the
+        # required count
+        if newspec is None:
+            self.log.info("no standbys affected for fs '{}'".format(fs_name))
+            return
+
+        self.log.info("new placement count:{0}".format(newspec.placement.count))
+
+        completion = self.apply_mds(newspec)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def notify(self, notify_type, notify_id):
+        # we don't know for which fs config has been changed
+        if notify_type != 'fs_map':
+            return
+        self.fs_map = self.get('fs_map')
+        if not self.fs_map:
+            return
+        for fs in self.fs_map.get('filesystems'):
+            fs_name = fs.get('mdsmap').get('fs_name')
```

Comment:
Use `fs.get('fscid')` instead of `fs_name`.

Reply:
all the operations done by the orchestrator are by matching the `fs_name`
operations local to mds_autoscaler module could be done by matching with the fscid

btw, is the `id` field in the mdsmap of a file-system as seen in `fs dump` the same as `fscid` that you've mentioned ?

Reply:
The problem with `mydict.get('key1').get('key2')` is the following: If `mydict` doesn't have the key `key1`, `get()` returns None, which doesn't have a `get()` method. So

1. either use `mydict['key1'].get(`key1`)`
2. or use a proper default, like `mydict.get('key1', {}).get('key2')`

Reply:
@batrick this is now fixed